[GFC] Initial implementation of grid and track sizing with fixed sized tracks.

sammygill opened the pull request:
#### 72484044d578b61dd47382f9a28c0376ea3015dd
<pre>
[GFC] Initial implementation of grid and track sizing with fixed sized tracks.
<a href="https://bugs.webkit.org/show_bug.cgi?id=299406">https://bugs.webkit.org/show_bug.cgi?id=299406</a>
<a href="https://rdar.apple.com/problem/161210933">rdar://problem/161210933</a>

Reviewed by Alan Baradlay.

Adds the initial support and implementation of track sizing with a focus
on fixed-size columns and rows. Like the patches before, we provide the
overall architecture that will be fleshed out over a series of patches.

At a high level, this code should eventually fully implement the &quot;Grid
Sizing Algorithm&quot; and &quot;Track Sizing Algorithm,&quot; portions of the spec:
<a href="https://drafts.csswg.org/css-grid-1/#algo-grid-sizing">https://drafts.csswg.org/css-grid-1/#algo-grid-sizing</a>
<a href="https://drafts.csswg.org/css-grid-1/#algo-track-sizing">https://drafts.csswg.org/css-grid-1/#algo-track-sizing</a>

For simplicity, we only implement steps one and two of the grid sizing
algorithm: first pass of column sizing and first pass of row sizing.
Since we are dealing with fixed-size tracks, we do not need to worry
about the available space portion of the text.

As input to this, we need the PlacedGridItems along with the track sizing
functions for each track. We output the used track sizes that come as a
result of the steps in the algorithm. Needless to say, this input and
output may need to change as we consider the more complicated and
involved portions of the algorithm, but this seems sufficient for now.

One thing that may be notable is that for the most part, the spec does
not distinguish between columns and rows in the steps of the track
sizing algorithm (i.e. it operates on a set of tracks without regard
to the dimension), so we will attempt to do the same in the code.
One small exception to this is that depending on which step we
are in within the grid sizing algorithm, the available space for the track
sizing algorithm may change.

* Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp: Added.
(WebCore::Layout::TrackSizingAlgorithm::sizeTracks):
(WebCore::Layout::TrackSizingAlgorithm::initializeTrackSizes):
For fixed size tracks this is fairly straightforward since both the base
size and growth limit get set to the size specified for the track.

Canonical link: <a href="https://commits.webkit.org/300467@main">https://commits.webkit.org/300467@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f32f146aa298942380c281cf04f5d7e6e649f16

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122532 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42240 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32925 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129141 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74631 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/684df74b-c1b3-40ab-a233-f8c47c54ec6a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124408 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42958 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50833 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93143 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61857 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7c583a55-d424-4641-a9d2-91f219162563) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125484 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34260 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109706 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73790 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a82dadb9-2ddc-4ea0-8429-7ee9e996b3e4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33245 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27863 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72625 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103913 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28073 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131864 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49473 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37654 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101675 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49848 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105926 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101543 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25806 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46918 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25062 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46242 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49331 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55080 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48798 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52149 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50480 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -0,0 +1,129 @@
+            // A flexible sizing function
+            // Use an initial growth limit of infinity.
+            if (maxTrackSizingFunction.isContentSized() || maxTrackSizingFunction.isFlex())
+                return LayoutUnit::max();
```
be careful with special values.

sammygill committed:
[GFC] Initial implementation of grid and track sizing with fixed sized tracks.
https://bugs.webkit.org/show_bug.cgi?id=299406
rdar://problem/161210933

Reviewed by Alan Baradlay.

Adds the initial support and implementation of track sizing with a focus
on fixed-size columns and rows. Like the patches before, we provide the
overall architecture that will be fleshed out over a series of patches.

At a high level, this code should eventually fully implement the "Grid
Sizing Algorithm" and "Track Sizing Algorithm," portions of the spec:
https://drafts.csswg.org/css-grid-1/#algo-grid-sizing
https://drafts.csswg.org/css-grid-1/#algo-track-sizing

For simplicity, we only implement steps one and two of the grid sizing
algorithm: first pass of column sizing and first pass of row sizing.
Since we are dealing with fixed-size tracks, we do not need to worry
about the available space portion of the text.

As input to this, we need the PlacedGridItems along with the track sizing
functions for each track. We output the used track sizes that come as a
result of the steps in the algorithm. Needless to say, this input and
output may need to change as we consider the more complicated and
involved portions of the algorithm, but this seems sufficient for now.

One thing that may be notable is that for the most part, the spec does
not distinguish between columns and rows in the steps of the track
sizing algorithm (i.e. it operates on a set of tracks without regard
to the dimension), so we will attempt to do the same in the code.
One small exception to this is that depending on which step we
are in within the grid sizing algorithm, the available space for the track
sizing algorithm may change.

* Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp: Added.
(WebCore::Layout::TrackSizingAlgorithm::sizeTracks):
(WebCore::Layout::TrackSizingAlgorithm::initializeTrackSizes):
For fixed size tracks this is fairly straightforward since both the base
size and growth limit get set to the size specified for the track.

Canonical link: https://commits.webkit.org/300467@main
